services/webnn/ort/graph_builder_ort.cc

```diff
@@ -733,6 +736,146 @@ GraphBuilderOrt::AddInstanceNormalizationOperation(
+  std::vector<const char*> input_names = {input_name.c_str()};
+  const std::vector<uint32_t>& input_shape =
+      GetOperand(layer_normalization.input_operand_id).descriptor.shape();
+
+  const std::string node_name = GetNodeName(layer_normalization.label);
+  const std::string output_name =
+      GetOperandName(layer_normalization.output_operand_id);
+  std::array<const char*, 1> output_names = {output_name.c_str()};
+
+  auto axes = layer_normalization.axes;
+  // ONNX doesn't support empty axes. When axes is empty, the mean equals to
+  // input, output = bias + (scale * 0)
+  if (axes.empty()) {
+    if (layer_normalization.bias_operand_id) {
+      base::CheckedNumeric<uint32_t> checked_input_size =
+          std::accumulate(input_shape.begin(), input_shape.end(),
+                          base::CheckedNumeric<uint32_t>(1), std::multiplies());
+      if (!checked_input_size.IsValid()) {
+        return NewNotSupportedError("The size of input is too large.");
+      }
+      std::vector<float> zero_data(checked_input_size.ValueOrDie(), 0.0f);
+      std::string zero_name;
+      switch (input_data_type) {
+        case OperandDataType::kFloat16: {
+          std::vector<uint16_t> zero_data_fp16 =
+              ConvertFloat32ToFloat16(zero_data);
+          zero_name = CreateInitializer<uint16_t>(input_shape, zero_data_fp16,
+                                                  input_data_type);
+          break;
+        }
+        case OperandDataType::kFloat32: {
+          zero_name =
+              CreateInitializer<float>(input_shape, zero_data, input_data_type);
+          break;
+        }
+        default:
+          NOTREACHED() << "[WebNN] InstanceNormalization only supports float32 "
+                          "and float16 data type.";
+      }
+      const std::string bias_name =
+          GetOperandName(layer_normalization.bias_operand_id.value());
+      std::array<const char*, 2> binary_input_names = {bias_name.c_str(),
+                                                       zero_name.c_str()};
+      model_builder_.AddNode(kOpTypeAdd, node_name, binary_input_names,
+                             output_names);
+    } else {
+      std::array<const char*, 2> binary_input_names = {input_name.c_str(),
+                                                       input_name.c_str()};
+      model_builder_.AddNode(kOpTypeSub, node_name, binary_input_names,
+                             output_names);
+    }
+    return base::ok();
+  }
+
+  // TODO: crbug.com/356905058: Figure out if unordered axes should be allowed.
+  if (!base::ranges::is_sorted(axes)) {
+    return NewNotSupportedError("Axes must be ordered for layerNormalization.");
+  }
+  const auto axes_size = axes.size();
+  // Here we only check beginning and ending of the ascending sorted axes,
+  // because the blink validation code ensures axes not having duplicated
+  // values.
+  // TODO: support inconsecutive axes by emulation.
```

Comment:
Please open an issue and update this TODO: https://github.com/shiyi9801/chromium/issues/new